### ISSUE TITLE
Type inference fails for custom type broadcasting

### ISSUE THREAD

**Issue by tansongchen** [FluxML/Zygote.jl#1359]
Let's say I have a broadcast over an array of custom type. The following MWE warns that the return type of `pb(d_result)` cannot be determined:

```julia
using Zygote, InteractiveUtils
import Base: exp

struct MyFloat64 <: Number
    n::Float64
end

exp(f::MyFloat64) = MyFloat64(exp(f.n))
my_vector = MyFloat64[1., 2., 3.]
result, pb = Zygote._pullback(broadcast, exp, my_vector)
d_result = MyFloat64[1., 1., 1.]
@code_warntype pb(d_result)
```

output is

```
MethodInstance for (::Zygote.var"#3909#back#952"{typeof(∂(broadcasted))})(::Vector{MyFloat64})
  from (::Zygote.var"#3909#back#952")(Δ) in Zygote at /Users/tansongchen/.julia/packages/ZygoteRules/AIbCs/src/adjoint.jl:67
Arguments
  #self#::Zygote.var"#3909#back#952"{typeof(∂(broadcasted))}
  Δ::Vector{MyFloat64}
Body::Tuple{Nothing, Any, Any}
1 ─ %1 = Core.getfield(#self#, Symbol("#3908#_back"))::typeof(∂(broadcasted))
│   %2 = ZygoteRules.unthunk_tangent(Δ)::Vector{MyFloat64}
│   %3 = (%1)(%2)::Tuple{Nothing, Any, Any}
│   %4 = (ZygoteRules.gradtuple0)(%3)::Tuple{Nothing, Any, Any}
└──      return %4
```

To my best knowledge of Zygote, the magic happens [here](https://github.com/FluxML/Zygote.jl/blob/master/src/lib/broadcast.jl#L204), but I couldn't tell what happens in the pullback function so that it's type unstable. `Real` and `Complex` are rescued by ForwardDiff, but not for custom types.

p.s. Found a similar issue #885 which is two years ago, but it didn't really get solved.

**Comment by ToucheSir** [comment]
Cthulhu reports a `Union{}` inside https://github.com/FluxML/Zygote.jl/blob/v0.6.54/src/lib/broadcast.jl#L205-L209, and running the pullback shows why:
```julia-repl
julia> pb(d_result)
ERROR: MethodError: no method matching conj(::MyFloat64)
Closest candidates are:
  conj(::Union{LinearAlgebra.Hermitian{T, S}, LinearAlgebra.Symmetric{T, S}} where {T, S}) at ~/.julia/juliaup/julia-1.8.5+0.x64.linux.gnu/share/julia/stdlib/v1.8/LinearAlgebra/src/symmetric.jl:368
  conj(::SparseArrays.SparseVector{<:Complex}) at ~/.julia/juliaup/julia-1.8.5+0.x64.linux.gnu/share/julia/stdlib/v1.8/SparseArrays/src/sparsevector.jl:1215
  conj(::ChainRulesCore.Tangent) at ~/.julia/packages/ChainRulesCore/C73ay/src/tangent_types/tangent.jl:167
  ...
Stacktrace:
  [1] exp_pullback
    @ ~/.julia/packages/ChainRules/RZYEu/src/rulesets/Base/fastmath_able.jl:56 [inlined]
  [2] ZBack
    @ ~/.julia/packages/Zygote/AS0Go/src/compiler/chainrules.jl:206 [inlined]
  [3] (::Zygote.var"#938#943")(::Tuple{MyFloat64, Zygote.ZBack{ChainRules.var"#exp_pullback#1313"{MyFloat64, ChainRulesCore.ProjectTo{Number, NamedTuple{(), Tuple{}}}}}}, ȳ₁::MyFloat64)
    @ Zygote ~/.julia/packages/Zygote/AS0Go/src/lib/broadcast.jl:205
  [4] (::Base.var"#4#5"{Zygote.var"#938#943"})(a::Tuple{Tuple{MyFloat64, Zygote.ZBack{ChainRules.var"#exp_pullback#1313"{MyFloat64, ChainRulesCore.ProjectTo{Number, NamedTuple{(), Tuple{}}}}}}, MyFloat64})
    @ Base ./generator.jl:36
  [5] iterate
    @ ./generator.jl:47 [inlined]
  [6] collect
    @ ./array.jl:787 [inlined]
  [7] map
    @ ./abstractarray.jl:3055 [inlined]
  [8] (::Zygote.var"#∇broadcasted#942"{Tuple{Vector{MyFloat64}}, Vector{Tuple{MyFloat64, Zygote.ZBack{ChainRules.var"#exp_pullback#1313"{MyFloat64, ChainRulesCore.ProjectTo{Number, NamedTuple{(), Tuple{}}}}}}}, Val{2}})(ȳ::Vector{MyFloat64})
    @ Zygote ~/.julia/packages/Zygote/AS0Go/src/lib/broadcast.jl:205
  [9] #3885#back
    @ ~/.julia/packages/ZygoteRules/AIbCs/src/adjoint.jl:67 [inlined]
 [10] #208
    @ ~/.julia/packages/Zygote/AS0Go/src/lib/lib.jl:206 [inlined]
 [11] #2066#back
    @ ~/.julia/packages/ZygoteRules/AIbCs/src/adjoint.jl:67 [inlined]
 [12] Pullback
    @ ./broadcast.jl:1298 [inlined]
 [13] (::Zygote.var"#3909#back#952"{typeof(∂(broadcasted))})(Δ::Vector{MyFloat64})
    @ Zygote ~/.julia/packages/ZygoteRules/AIbCs/src/adjoint.jl:67
 [14] top-level scope
    @ REPL[11]:1
```

**Comment by ToucheSir** [comment]
I also found the correct methods to implement eventually. The issue appears to be some compiler heuristic being violated, which leads to no const prop from ntuple and `StaticGetter` not being so static. PR in #1360.

**Comment by CarloLucibello** [comment]
Closed by #1360
